lib/rust/ensogl/component/drop-down/src/lib.rs

```diff
@@ -0,0 +1,272 @@
+
+            output.entries_in_range_needed <+ requested_range_needed;
+
+            visible_range <- model.grid.viewport.map(|viewport| {
+                let start = (-viewport.top / ENTRY_HEIGHT).floor() as usize;
+                let end = (-viewport.bottom / ENTRY_HEIGHT).ceil() as usize;
+                start..end
+            });
+            output.currently_visible_range <+ visible_range;
+
+            updated_range <- provided_entries.map4(
+                &visible_range, &max_cache_size, &number_of_entries,
+                f!(((range, entries), visible, cache_size, num_entries)
+                    model.insert_entries_in_range(range.clone(), entries, visible.clone(), *cache_size, *num_entries))
+            );
+
+            range_to_update <- any(updated_range, requested_range_ready);
+            model.grid.model_for_entry <+ range_to_update.map(f!([model](range) {
+                let models_with_index = model.entry_models_for_range(range.clone());
+                let models_with_cell_pos = models_with_index.map(|(idx, entry)| (idx, 0, entry));
+                models_with_cell_pos.collect_vec()
+            })).iter();
+
+
+            // === Selection ===
+            selection_pruned <- input.set_multiselect.map(f!((multi) model.set_multiselect(*multi))).on_true();
+            selection_accepted <- model.grid.entry_accepted.map3(
+                &input.set_multiselect, &input.allow_deselect_all,
+                f!(((row, _), multi, allow) model.accept_entry_at_index(*row, *multi, *allow)));
+            selection_set <- input.set_selected_entries.map2(&input.set_multiselect,
+                f!((values, max) model.set_selection(values, *max)));
+            selection_changed <- any3(&selection_accepted, &selection_set, &selection_pruned);
+
+            model.grid.request_model_for_visible_entries <+ selection_changed;
+            output.selected_entries <+ selection_changed.map(f!((()) model.get_selected_entries())).on_change();
+            output.single_selected_entry <+ selection_changed.map(f!((()) model.get_single_selected_entry())).on_change();
+
+
+            // === Keyboard navigation ===
+            model.grid.accept_selected_entry <+ input.toggle_focused_entry;
+            model.grid.move_selection_up <+ input.focus_previous_entry;
+            model.grid.move_selection_down <+ input.focus_next_entry;
+            model.grid.select_entry <+ model.grid.entry_hovered;
+
+            has_focused_entry <- model.grid.entry_selected.map(|entry| entry.is_some());
+            model.grid.select_entry <+ input.focus_previous_entry.gate_not(&has_focused_entry)
+                .map2(&visible_range, |_, range| Some(((range.end - 1).max(range.start), 0)));
+            model.grid.select_entry <+ input.focus_next_entry.gate_not(&has_focused_entry)
+                .map2(&visible_range, |_, range| Some((range.start, 0)));
+
+
+            // === Initialization ===
```

Comment:
Please refactor these FRP inits to several init functions.

Reply:
I tried to do that now, but it turns out to be massively verbose. Because the `component::Frp` trait doesn't use `self` value for initialization, each init function needs to receive network, private frp Api and model as separate arguments. Then, some sections use streams defined in previous sections (on purpose). Spitting it into multiple methods also means that I need to weave those streams through returns and extra params.

I think leaving it as a single FRP initialization block is much easier to follow and it is definitely less verbose.

Reply:
I'm fine to change that. I'll need to understand the reasons behind original design of it, but in general I agree that it should change to allow easier code splitting. Let's do that as part of future widget PRs.